### PR TITLE
support private connection requests

### PR DESCRIPTION
**Pull request by naorzr**
as part of #68

### CODE REVIEW

File: tunnel-server/index.ts [hunk 1]
```diff
@@ -55,10 +50,10 @@ const sshServer = createSshServer({
   log: sshLogger,
   sshPrivateKey,
   socketDir: '/tmp', // TODO
-  onPipeCreated: async (clientId, remotePath, localSocket) => {
+  onPipeCreated: async ({clientId, remotePath, localSocketPath, publicKey, access}) => {
```
Comment (Yshayy) [inline tunnel-server/index.ts:53]: I think that for consistency, both onPipeCreated and onPipeDestroyed should use props style signature

File: tunnel-server/src/proxy.ts [hunk 2]
```diff
@@ -52,6 +58,24 @@ export function proxyHandlers({
         return;
       }
 
+      if(env.access === 'private'){
+        const authorization = req.headers['authorization'];
+        if (!authorization) {
+          return unauthorized(res)
+        }
+
+        const basicAuth = Buffer.from(authorization?.split('Basic ')[1] || '', 'base64').toString('ascii');
```
Comment (Yshayy) [inline tunnel-server/src/proxy.ts:67]: Might worth splitting the authentication scheme and check if it's basic.
http://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml

File: tunnel-server/src/proxy.ts [hunk 3]
```diff
@@ -52,6 +58,24 @@ export function proxyHandlers({
...
+        const [username, password] = basicAuth.split(':');
+        if (username !== 'x-preevy-key'){
```
Comment (Yshayy) [inline tunnel-server/src/proxy.ts:69]: Can be `x-preevy-profile-key` to be more specific

File: tunnel-server/src/proxy.ts [hunk 4]
```diff
@@ -52,6 +58,24 @@ export function proxyHandlers({
...
+          return unauthorized(res)
+        }
+        const signature = Buffer.from(password, 'base64');
```
Comment (Yshayy) [inline tunnel-server/src/proxy.ts:72]: If the data is already wrapped in base64, is there any text format that can be stored binary data more efficiently and still passed in clear text and in url?

**Review by Yshayy — COMMENTED** [review]
LGTM :)
Added few notes

File: tunnel-server/src/proxy.ts (hunk 4 above)
Comment (naorzr) [reply tunnel-server/src/proxy.ts:72]: for arbitrary binary data base64 is the most space efficient

File: tunnel-server/src/proxy.ts [hunk 5]
```diff
@@ -52,6 +58,24 @@ export function proxyHandlers({
         return;
       }
 
+      if(env.access === 'private'){
```
Comment (royra) [inline tunnel-server/src/proxy.ts:61]: extract to function?

File: tunnel-server/src/ssh-server.ts [hunk 6]
```diff
@@ -2,12 +2,26 @@ import crypto, { randomBytes } from 'crypto'
...
 import path from 'path'
-import ssh2 from 'ssh2'
+import ssh2, { ParsedKey, SocketBindInfo } from 'ssh2'
 import { inspect } from 'util'
 
 const idFromPublicSsh = (key: Buffer) =>
   crypto.createHash('sha1').update(key).digest('base64url').replace(/[_-]/g, '').slice(0, 8).toLowerCase()
 
+const getRequestedSocketPath = (info: ssh2.SocketBindInfo) => {
+  const [path, access] = info.socketPath.split('#')
```
Comment (royra) [inline tunnel-server/src/ssh-server.ts:12]: I would choose an extendable format like
`mypath;auth=profile-key`
this way we can add more stuff here

File: tunnel-server/src/proxy.ts [hunk 7]
```diff
@@ -52,6 +58,24 @@ export function proxyHandlers({
...
+        const authorization = req.headers['authorization'];
+        if (!(authorization && authorization.startsWith('Basic '))) {
+          return unauthorized(res)
+        }
+
+        const basicAuth = Buffer.from(authorization?.split('Basic ')[1] || '', 'base64').toString('ascii');
+        const [username, password] = basicAuth.split(':');
+        if (username !== 'x-preevy-profile-key'){
+          return unauthorized(res)
```
Comment (royra) [inline tunnel-server/src/proxy.ts:70]: add log

File: tunnel-server/src/proxy.ts [hunk 8]
```diff
@@ -52,6 +58,24 @@ export function proxyHandlers({
...
 
+      if(env.access === 'private'){
+        const authorization = req.headers['authorization'];
+        if (!(authorization && authorization.startsWith('Basic '))) {
+          return unauthorized(res)
+        }
+
+        const basicAuth = Buffer.from(authorization?.split('Basic ')[1] || '', 'base64').toString('ascii');
+        const [username, password] = basicAuth.split(':');
+        if (username !== 'x-preevy-profile-key'){
```
Comment (royra) [inline tunnel-server/src/proxy.ts:69]: maybe the username should be the clientId?

**Pull request merged**